File: app/models/resume.rb
# frozen_string_literal: true

# == Schema Information
#
# Table name: resumes
#
#  id                     :integer          not null, primary key
#  about_myself           :text
#  answers_count          :integer          default(0), not null
#  awards_description     :text
#  city                   :string
#  contact                :string
#  contact_email          :string
#  contact_phone          :string
#  contact_telegram       :string
#  english_fluency        :string
#  evaluated_ai           :boolean
#  evaluated_ai_state     :string
#  github_url             :string
#  hexlet_url             :string
#  impressions_count      :integer          default(0)
#  locale                 :string
#  name                   :string           not null
#  projects_description   :text
#  relocation             :string
#  skills_description     :text(250)
#  skills_description_old :text
#  state                  :string
#  summary                :text
#  url                    :string
#  created_at             :datetime         not null
#  updated_at             :datetime         not null
#  user_id                :integer          not null
#
# Indexes
#
#  index_resumes_on_user_id  (user_id)
#
# Foreign Keys
#
#  user_id  (user_id => users.id)
#
class Resume < ApplicationRecord
  include StateConcern
  extend Enumerize
  extend TagResumePresenter

  mark_as_outdated :hexlet_url, :awards_description

  has_paper_trail
  is_impressionable counter_cache: true

  acts_as_taggable_on :skills, :directions

  enumerize :english_fluency, in: %i[dont_know basic read pass_interview fluent], scope: true, predicates: { prefix: true }
  enumerize :locale, in: I18n.available_locales
  enumerize :relocation, in: %i[not_specified another_country another_city another_city_country not_ready], scope: true, predicates: { prefix: true }

  validates :name, presence: true
  validates :github_url, presence: true, unless: :draft?
  validates :summary, length: { minimum: 200 }, presence: true, unless: :draft?
  validates :skills_description, length: { maximum: 250 }, presence: true, unless: :draft?
  validates :skills_description, presence: true, unless: :draft?
  validates :contact_email, presence: true, unless: :draft?
  validates :contact_email, 'valid_email_2/email': true
  validates :contact_phone, phone: true

  belongs_to :user
  has_many :answers, inverse_of: :resume, dependent: :destroy
  has_many :answer_likes, through: :answers, source: :likes
  has_many :educations, inverse_of: :resume, dependent: :destroy
  has_many :works, inverse_of: :resume, dependent: :destroy
  has_many :comments, inverse_of: :resume, dependent: :destroy

  include ResumeRepository

  accepts_nested_attributes_for :educations, reject_if: :all_blank, allow_destroy: true
  accepts_nested_attributes_for :works, allow_destroy: true

  ransack_alias :popular, :impressions_created_at_or_comments_created_at_or_answers_created_at

  aasm :state, column: :state do
    state :draft, initial: true
    state :published
    state :archived

    event :publish do
      transitions from: %i[draft published], to: :published
    end

    event :hide do
      transitions from: %i[draft published], to: :draft
    end

    event :archive do
      transitions from: %i[published], to: :archived
    end

    event :restore do
      transitions from: %i[archived], to: :published
    end
  end

  aasm :evaluated_ai_state, column: :evaluated_ai_state do
    state :not_evaluated, initial: true
    state :processing
    state :evaluated
    state :failed

    event :process do
      transitions from: %i[not_evaluated failed], to: :processing, guard: :published?
    end

    event :mark_as_evaluated do
      transitions from: :processing, to: :evaluated
    end

    event :mark_as_failed do
      transitions from: :processing, to: :failed
    end
  end

  def initialize(attribute = nil)
    defaults = {
      locale: I18n.locale,
      relocation: :not_specified
    }

    attrs_with_defaults = attribute ? defaults.merge(attribute) : defaults
    super(attrs_with_defaults)
  end

  def to_s
    name
  end

  def author?(user)
    user_id == user.id
  end

  def self.ransackable_attributes(_auth_object = nil)
    %w[id answers_count created_at impressions_count name state]
  end

  def self.ransackable_associations(_auth_object = nil)
    %w[directions user skills answers]
  end
end

File: src/main/java/io/hexlet/cv/model/Resume.java
package io.hexlet.cv.model;

import static jakarta.persistence.GenerationType.IDENTITY;

import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EntityListeners;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.Id;
import jakarta.persistence.Index;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToMany;
import jakarta.persistence.Table;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

@Entity
@Table(name = "resumes", indexes = {
    @Index(name = "idx_resumes_user_id", columnList = "user_id")
})
@Getter
@Setter
@NoArgsConstructor
@EntityListeners(AuditingEntityListener.class)
public class Resume {
    @Id @GeneratedValue(strategy = IDENTITY)
    private Long id;

    private String state;

    @Column(nullable = false)
    private String name;

    @ManyToOne(optional = false, fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    private String url;

    @Column(columnDefinition = "text")
    private String summary;

    @Column(columnDefinition = "text")
    private String skillsDescription;

    private String githubUrl;

    @Column(columnDefinition = "text")
    private String awardsDescription;

    private String englishFluency;
    private Integer impressionsCount;

    @Column(nullable = false)
    private Integer answersCount;

    private String hexletUrl;
    private String contact;
    private String locale;
    private String city;
    private String relocation;
    private String contactPhone;
    private String contactEmail;
    private String contactTelegram;
    private Boolean evaluatedAi;
    private String evaluatedAiState;

    @Column(columnDefinition = "text")
    private String projectsDescription;

    @Column(columnDefinition = "text")
    private String aboutMyself;

    @CreatedDate
    private LocalDateTime createdAt;

    @LastModifiedDate
    private LocalDateTime updatedAt;


    @OneToMany(mappedBy = "resume", cascade = CascadeType.REMOVE, orphanRemoval = true)
    private List<ResumeAnswer> answers = new ArrayList<>();

    @OneToMany(mappedBy = "resume", cascade = CascadeType.REMOVE, orphanRemoval = true)
    private List<ResumeComment> comments = new ArrayList<>();

    @OneToMany(mappedBy = "resume", cascade = CascadeType.REMOVE, orphanRemoval = true)
    private List<ResumeEducation> educations = new ArrayList<>();

    @OneToMany(mappedBy = "resume", cascade = CascadeType.REMOVE, orphanRemoval = true)
    private List<ResumeWork> works = new ArrayList<>();


}
